File: Python_HW/엄성민/hw_c++/zodiac.cpp
#include <iostream>
#include <sstream>
#include <vector>
#include <algorithm>


using namespace std;
bool containsNumber(const string& str) {
    for (char c : str) {
        if (!(isdigit(c) or ' ')) {
            return false;  
        }
    }
    return true;  
}

void count_year(int year,int* list ) {
    int a = year % 12;
    list[a]++;
}

int main() {

    string input;
    
    cout << "숫자들을 공백으로 구분하여 입력하세요: ";
    getline(cin, input);
    while (!containsNumber(input)) {
        getline(cin, input);
        cout << "숫자와 공백만 입력하세요" << endl;
    }

    string animal_list[12] = { "Monkey","Rooster","Dog","Pig","RAt","Ox","Tigger",
        "Rabbit","Dragon","snake","Horse","Sheep" };
    vector<int> numbers; 

    istringstream iss(input); 
    int count[12] = { 0 };
    int num;
    while (iss >> num) { 
        numbers.push_back(num);  
    }
    int total_number = 0;

    for (int num : numbers) {
        count_year(num, count);
        //숫자들 함수에 넣어서 나눠서 확인
    }
    for (int i = 0; i < 12; i++) {
        if (count[i] != 0) {
            total_number++;
        }
    }
    auto maxIt = max_element(numbers.begin(), numbers.end());
    int maxValue = *maxIt;
    int k = maxValue % 12;

    cout << total_number << "  " << animal_list[k];


    return 0;
}//finish

File: Python_HW/서지민/zodiac.py
def find_earliest_zodiac(years):
    zodiac_order = {
        "Rat": 0,
        "Ox": 1,
        "Tiger": 2,
        "Rabbit": 3,
        "Dragon": 4,
        "Snake": 5,
        "Horse": 6,
        "Sheep": 7,
        "Monkey": 8,
        "Rooster": 9,
        "Dog": 10,
        "Pig": 11
    }

    zodiac_counts = {}
    earliest_zodiac = "Pig"

    for year in years:
        zodiac = (year - 4) % 12
        zodiac_name = list(zodiac_order.keys())[list(zodiac_order.values()).index(zodiac)]

        if zodiac_name not in zodiac_counts:
            zodiac_counts[zodiac_name] = 1
        else:
            zodiac_counts[zodiac_name] += 1

        if zodiac_order[zodiac_name] < zodiac_order[earliest_zodiac]:
            earliest_zodiac = zodiac_name

    num_different_zodiacs = len(zodiac_counts)

    return num_different_zodiacs, earliest_zodiac

# 입력 받기
input_years = input().split()
lunar_years = [int(year) for year in input_years]

# 결과 출력
num_zodiacs, earliest_zodiac = find_earliest_zodiac(lunar_years)
print(num_zodiacs, earliest_zodiac)
